# Check Fitz Overlap with SkinCAP

This Jupyter Notebook is for checking the overlap between the Fitzpatrick images we have and the SkinCAP Fitzpatrick images we have.

In [20]:
from google.cloud import storage
import pandas as pd
from io import StringIO, BytesIO

pd.set_option("display.max_rows", None)
pd.set_option("display.max_columns", None)

# load SkinCAP images
client = storage.Client()
bucket = client.bucket('derma-datasets-2')

def load_csv(path):
    blob = bucket.blob(path)
    data = blob.download_as_text()
    csv = pd.read_csv(StringIO(data))
    return csv

# skincap_csv_blob = bucket.blob('raw/SkinCAP/skincap_v240623.csv')
# skincap_csv_data = skincap_csv_blob.download_as_text()
# skincap_meta = pd.read_csv(StringIO(skincap_csv_data))

skincap_meta = load_csv('raw/SkinCAP/skincap_v240623.csv')

In [22]:
skincap_meta = skincap_meta[["id", "skincap_file_path", "ori_file_path", "source"]]
skincap_meta.head()

id skincap_file_path ori_file_path source
0   1             1.png    000002.png    ddi
1   2             2.png    000003.png    ddi
2   3             3.png    000004.png    ddi
3   4             4.png    000005.png    ddi
4   5             5.png    000006.png    ddi

In [3]:
skincap_meta['source'].isna().any()

np.False_

In [4]:
skincap_meta_fitz_flg = skincap_meta["source"] == "fitzpatrick17k"
skincap_meta_fitz = skincap_meta[skincap_meta_fitz_flg]

In [5]:
skincap_meta_fitz.head()

id skincap_file_path                         ori_file_path  \
655  656           656.png  fa2911a9b13b6f8af79cb700937cc14f.jpg   
656  657           657.png  e702b1a7dc40aa1d8e85ccdb019c4ab2.jpg   
657  658           658.png  8438db40abd1eccfbc7ee4b469f1b6f1.jpg   
658  659           659.png  9a3af1bc39e115bcc6931170cf8a00bb.jpg   
659  660           660.png  59ccb668671950ca657b6bc48213b763.jpg   

             source  
655  fitzpatrick17k  
656  fitzpatrick17k  
657  fitzpatrick17k  
658  fitzpatrick17k  
659  fitzpatrick17k

In [6]:
# filter SkinCAP data to only consider Fitzpatrick images not in final Fitzpatrick dataset (i.e. the image file is not in the raw Fitzpatrick bucket)

def list_files(prefix, get_full_path=False):
    if not prefix.endswith("/"): prefix = prefix + "/"
    
    blobs = bucket.list_blobs(prefix=prefix)
    filepaths = []
    for blob in blobs:
        filepaths.append(blob.name)

    if get_full_path:
        return filepaths

    filenames = [path.replace(prefix, "") for path in filepaths]
    return filenames

fitz_img_filenames = list_files("raw/fitzpatrick17k/images")
fitz_img_filenames[0:5]

['000e8dd5ee75dd6668e978e7a4e6fe54.jpg',
 '001b1edd9c7a02e9381793f6e9081d20.jpg',
 '002f4ec0b620339d58ccfa26c4268584.jpg',
 '005c26fd71696854f7b4c9d80a5dc9d6.jpg',
 '0066264501f73b1db09b3e2aa4dac0a0.jpg']

In [7]:
# filter SkinCAP dataset to only include FitzPatrick images that we have not downloaded
not_in_fitz_flg = ~skincap_meta_fitz['ori_file_path'].isin(fitz_img_filenames)
skincap_dont_have = skincap_meta_fitz[not_in_fitz_flg]
skincap_dont_have.shape

(2811, 4)

In [8]:
skincap_dont_have.head()

id skincap_file_path                         ori_file_path  \
655  656           656.png  fa2911a9b13b6f8af79cb700937cc14f.jpg   
656  657           657.png  e702b1a7dc40aa1d8e85ccdb019c4ab2.jpg   
657  658           658.png  8438db40abd1eccfbc7ee4b469f1b6f1.jpg   
658  659           659.png  9a3af1bc39e115bcc6931170cf8a00bb.jpg   
659  660           660.png  59ccb668671950ca657b6bc48213b763.jpg   

             source  
655  fitzpatrick17k  
656  fitzpatrick17k  
657  fitzpatrick17k  
658  fitzpatrick17k  
659  fitzpatrick17k

In [9]:
# of these, which do we have in the SkinCAP dataset?

# get list of all the files we have in SkinCAP
sc_img_filenames = list_files("raw/SkinCAP/skincap/")
sc_img_filenames = [name for name in sc_img_filenames if name.endswith("png")]
print(len(sc_img_filenames))

sc_img_filenames += [name for name in list_files("raw/SkinCAP/skincap/not_include") if name.endswith("png")]
sc_img_filenames = list(set(sc_img_filenames)) # drop duplicates
print(len(sc_img_filenames))

4346
4692


In [10]:
# check to make sure we actually have SkinCAP files for all the files in skincap_dont_have
not_in_either_flg = ~skincap_dont_have["skincap_file_path"].isin(sc_img_filenames)
skincap_dont_have[not_in_either_flg]

Empty DataFrame
Columns: [id, skincap_file_path, ori_file_path, source]
Index: []

Of the images we have in SkinCAP, abotu 2811 come from the Fitzpatrick, but we were not able to download them. This only includes the files in the main `skincap` folder. Let's see if any of the files in the `not_include` folder are also part of Fitzpatrick. We can do that by taking the files, computing the hash, and checking if it is in the full Fitzpatrick dataset.

In [13]:
import hashlib
from concurrent.futures import ThreadPoolExecutor, as_completed

def compute_md5_for_blob(blob):
    md5 = hashlib.md5()
    with blob.open("rb") as f:
        while True:
            data = f.read(8192)
            if not data:
                break
            md5.update(data)
    return blob.name, md5.hexdigest()

def compute_md5_gcs_folder_parallel(bucket_name, prefix, max_workers=16):
    blobs = [
        blob for blob in client.list_blobs(bucket_name, prefix=prefix)
        if blob.name.lower().endswith(('.jpg', '.jpeg', '.png', '.gif', '.bmp', '.webp', '.tiff'))
    ]
    hashes = {}

    with ThreadPoolExecutor(max_workers=max_workers) as executor:
        future_to_blob = {executor.submit(compute_md5_for_blob, blob): blob for blob in blobs}
        for future in as_completed(future_to_blob):
            name, md5sum = future.result()
            hashes[name] = md5sum

    return hashes

# Usage:
bucket_name = "derma-datasets-2"
prefix      = "raw/SkinCAP/skincap/not_include/"
md5s = compute_md5_gcs_folder_parallel(bucket_name, prefix)
# for name, md5sum in md5s.items():
#     print(f"{name}: {md5sum}")

not_include_hashes = pd.DataFrame(list(md5s.items()), columns=["file_name", "md5_hash"])

In [14]:
not_include_hashes.head()

file_name                          md5_hash
0  raw/SkinCAP/skincap/not_include/4009.png  5ca94c02d89651613468462514c5d23b
1  raw/SkinCAP/skincap/not_include/4007.png  06967f5227a4227896b2b7c5e6901652
2  raw/SkinCAP/skincap/not_include/4002.png  60abb67a831bace6b777a6f4d82b7032
3  raw/SkinCAP/skincap/not_include/4003.png  c94e4af9862e5883872c85f4912433bd
4  raw/SkinCAP/skincap/not_include/4006.png  40c298a68522a302b8a22ebf3395194e

In [17]:
fitz_meta = load_csv('raw/fitzpatrick17k/fitzpatrick17k.csv')
fitz_meta.head()

md5hash  fitzpatrick_scale  fitzpatrick_centaur  \
0  5e82a45bc5d78bd24ae9202d194423f8                  3                    3   
1  fa2911a9b13b6f8af79cb700937cc14f                  1                    1   
2  d2bac3c9e4499032ca8e9b07c7d3bc40                  2                    3   
3  0a94359e7eaacd7178e06b2823777789                  1                    1   
4  a39ec3b1f22c08a421fa20535e037bba                  1                    1   

                             label nine_partition_label three_partition_label  \
0  drug induced pigmentary changes         inflammatory        non-neoplastic   
1                  photodermatoses         inflammatory        non-neoplastic   
2                   dermatofibroma        benign dermal                benign   
3                        psoriasis         inflammatory        non-neoplastic   
4                        psoriasis         inflammatory        non-neoplastic   

    qc                                                url  \
0  NaN  https://www.dermaamin.com/site/images/clinical...   
1  NaN  https://www.dermaamin.com/site/images/clinical...   
2  NaN  https://www.dermaamin.com/site/images/clinical...   
3  NaN  https://www.dermaamin.com/site/images/clinical...   
4  NaN  https://www.dermaamin.com/site/images/clinical...   

                                        url_alphanum  
0  httpwwwdermaamincomsiteimagesclinicalpicmminoc...  
1  httpwwwdermaamincomsiteimagesclinicalpicpphoto...  
2  httpwwwdermaamincomsiteimagesclinicalpicdderma...  
3  httpwwwdermaamincomsiteimagesclinicalpicppsori...  
4  httpwwwdermaamincomsiteimagesclinicalpicppsori...

In [19]:
fitz_meta_we_have = fitz_meta.merge(not_include_hashes, how="left", left_on="md5hash", right_on="md5_hash")
fitz_meta_we_have[fitz_meta_we_have["file_name"].notna()]

md5hash  fitzpatrick_scale  \
15053  50bb16c1bfa029e80a55ce64149f3338                  5   
15055  60abb67a831bace6b777a6f4d82b7032                  3   
15056  c94e4af9862e5883872c85f4912433bd                  3   
15058  331c09538efd0f63138152a5de9c1bbc                  4   
15059  20851c52ed096af52a69bb5a2670c505                  2   
15061  40c298a68522a302b8a22ebf3395194e                  4   
15067  06967f5227a4227896b2b7c5e6901652                  3   
15076  c3994c3b7fc8f7f537ce70972618c80d                  2   
15077  5ca94c02d89651613468462514c5d23b                  4   
15083  bb79039b52e9e693c47056248372931b                  2   
15087  3bf3afa6c145d67e06f6e839b1699a2d                  2   
15089  bdfcc02ca14c35a728b619018afbe816                  4   
15092  5b0fd3a791bce69a15553e3cec20ba87                  2   
15097  26b787b6e298c158fa14bc0edfbaafd9                  3   
15105  72c70306aace54ba95d9e65eae05f7eb                  5   
15107  e668628ce95135ab58957a41687fa659                  1   
15110  45a0207eae8a4305c1030a9ead023953                  3   
15111  20e54e08b45a31387075f7ecd46ae5fd                  3   
15114  13b2ea8d8b6e7ef3e4e3ea70e2078557                  4   
15119  6a2f26a865f05801a67349769382464a                  4   
15120  83167c5a26bd4a459c381834ecf97558                  1   
15127  5e53e934766adcc989d0f3e232d485dc                  5   
15131  98aec15c1b26da814698036a149c04f2                  5   
15135  fb847a2133594e00f8014338a75ab83c                 -1   
15143  dc84ce6b158006be88a8b1e4c0d86e99                  2   
15146  938528c2ba772fc83832aeca85171cf2                  1   
15157  dc411f7b9a5cec568f5865f6f3d5ca0a                  6   
15159  442efb92b5f59923668f0d2e6316b9cb                  4   
15167  a79a3adf796973f79dbe1795f6fa9f77                 -1   
15168  66617b242abb82dcd77cf73970510afa                  3   
15169  de24cc0b5dc20eccdd67e7a60c8a1fa9                  4   
15170  de48088daa1f726a293ad0e3ba2a3796                  5   
15172  7ad362076a87f9c248d9e43bdca031af                  5   
15183  1fe42da54970a24725ba4ee204bfab90                  5   
15188  06c312978164d6cc204751b5d7bea2f8                  3   
15195  01ba602def4506d8bfd5900cccd2ab4d                  2   
15196  4dbb4fad384acdcdb70f2fb191036b12                  4   
15201  130b528100181f8d69d1429f9cb20254                  2   
15203  0700fe5ecff021f968dc71700dbdf976                  1   
15204  0eae940a2c38375a082cc48bfc591676                  3   
15219  2131d014a301d4773e32bdc6ef466b53                  4   
15231  782b4fbbebe47fc5535187b65d0b96b7                  2   
15233  e96fab58eb38baf4993067b59ad327e4                  4   
15236  7189734192a159e6e8d153180e0f40e3                  3   
15239  670e2d45784467ce1ef243cd62bde464                  4   
15241  45310e88ea83bc2b728595b84abab12b                  4   
15244  f528a2d620dbe5e3348b3ff18a9f7d52                  2   
15247  688ac2ba9aff76f589c14be2813f9d5f                  4   
15250  76350c7ae10969705233d7fcc6e6e59c                  4   
15251  72425b50cd9e8658f118ff4156c5b3b5                  4   
15257  e2ec3ed300a3bae68e03b1a3f5992c5f                  5   
15259  127b90b0392a0eb74decc9ae0f526821                  2   
15269  6c83bef6f525b8734ecdbd9dd5f0e822                  3   
15271  324aca3793bba66f7c8ba6eea67c42e6                  4   
15278  8b53820171cc4ef6e17c56ec23c44e57                  3   
15284  0e6fb6688ac197411ca1146ed2018754                  2   
15291  9a9e24f5eeabe3cad0ddd6600bae7c0d                  1   
15304  230294f2e89a02e283b518e42450ea46                  3   
15311  eec6fc5623b1d46fa196daad1e92d3a5                  4   
15312  a37df08421905f79fc29feae76aca0f1                  6   
15316  3b0a85962bfea9a937d4619ef7e3ea37                  2   
15318  0e802991cd0257f69f62adedb41152f8                  3   
15319  b171e257fdc458f41480a167caf2b961                  2   
15321  aba1cb1aed55ede8db7ccdd9bcd5e9ff                  3   
15